src/Features/Core/Portable/Completion/Providers/ImportCompletionProvider/ExtensionMethodImportCompletionHelper.cs

```diff
@@ -265,197 +280,168 @@ async Task<(ImmutableArray<SerializableImportCompletionItem>, StatisticCounter)>
-                // types explicitly here.
-                //
-                // TODO:
-                // Alternatively, we can try to get symbols out of each assembly, instead of the compilation (which includes all references), 
-                // to avoid such conflict. We should give this approach a try and see if any perf improvement can be achieved.
-                var containerSymbol = compilation.GetTypeByMetadataName(fullyQualifiedContainerName);
-
-                if (containerSymbol != null)
+                var containerSymbol = assembly.GetTypeByMetadataName(fullyQualifiedContainerName);
+
+                if (containerSymbol == null
+                    || !containerSymbol.MightContainExtensionMethods
+                    || ShouldIgnoreSymbol(containerSymbol, ignoreInternal, ignoreUnspecifiedContainerAccessibility))
                 {
-                    GetItemsFromTypeContainsPotentialMatches(containerSymbol, qualifiedNamespaceName, methodNames, receiverTypeSymbol, semanticModel, position, counter, builder);
+                    continue;
                 }
-                else
+
+                foreach (var methodName in methodNames)
                 {
-                    conflictTypeRootNode.Add(fullyQualifiedContainerName, (qualifiedNamespaceName, methodNames));
+                    var methodSymbols = containerSymbol.GetMembers(methodName).OfType<IMethodSymbol>();
+
+                    foreach (var methodSymbol in methodSymbols)
+                    {
+                        counter.TotalExtensionMethodsChecked++;
+
+                        if (methodSymbol.IsExtensionMethod &&
+                            !ShouldIgnoreSymbol(methodSymbol, ignoreInternal, ignoreUnspecifiedMethodAccessibility))
+                        {
+                            // Find a potential match.
+                            builder.Add(methodSymbol);
+                        }
+                    }
                 }
             }
 
-            var ticks = Environment.TickCount;
-
-            GetItemsFromConflictingTypes(rootNamespaceSymbol, conflictTypeRootNode, builder, receiverTypeSymbol, semanticModel, position, counter);
-
-            counter.GetSymbolExtraTicks = Environment.TickCount - ticks;
-
             return builder.ToImmutable();
+
+            // A syntax based check for accessibility, a semantic based check is still required later.
+            // Since we are dealing with extension methods and their container (top level static class and modules),
+            // only public, internal and private modifiers are in play here.
+            static bool ShouldIgnoreSymbol(ISymbol symbol, bool ignoreInternal, bool ignoreUnspecifiedAccessibility)
+                => symbol.DeclaredAccessibility switch
+                {
+                    Accessibility.Public => false,
+                    Accessibility.Internal => ignoreInternal,
+                    Accessibility.NotApplicable => ignoreUnspecifiedAccessibility,
```

Comment:
i don't understand when the NotApplicable bit would happen.

Reply:
When they are declared w/o explicit accessibility modifier.

Reply:
I was wrong, fixed